File: packages/effects/src/filters.ts
// The optional spatial filters of the photograph pipeline (OPENERS.md, "Photograph pipeline"
// steps 2 and 3): a Gaussian blur before the fit, a minimum filter that thickens engraved
// hatching, and an unsharp mask applied to a band of rows after the fit.
//
// Pinned definitions (SPEC 5.4: one specified algorithm everywhere). These are the Turboslide
// definitions, not Pillow's: Pillow approximates its GaussianBlur with three extended box blurs
// and shrinks the image under a MinFilter. The Rust crate of SPEC 10 must reproduce the
// definitions below, and twins made by the Pillow rounds with a blur, minFilter or unsharp step
// may differ from the regenerated twins in a small share of cells; the parity test of M5 lists
// them (MILESTONES M5 acceptance, "missingParameter" residual).
//   gaussianBlur(radius): sigma = radius, kernel radius ceil(3 sigma), weights exp(-x^2 / 2 sigma^2)
//     normalized to one, separable, edges replicate, each pass rounded to the nearest integer.
//   minFilter(size): the minimum over the size by size window, edges replicate, same size out.
//   unsharpBand(rows, amount, radius, threshold): Pillow's UnsharpMask rule, out = in + (in -
//     blur) * amount / 100 where |in - blur| > threshold, on rows [from, to) only.
import type { GrayImage } from './image.ts';

function gaussianKernel(sigma: number): Float64Array {
  const radius = Math.ceil(3 * sigma);
  const k = new Float64Array(radius * 2 + 1);
  let sum = 0;
  for (let i = -radius; i <= radius; i += 1) {
    const w = Math.exp(-(i * i) / (2 * sigma * sigma));
    k[i + radius] = w;
    sum += w;
  }
  for (let i = 0; i < k.length; i += 1) k[i] = (k[i] ?? 0) / sum;
  return k;
}

function clampIndex(i: number, n: number): number {
  return i < 0 ? 0 : i >= n ? n - 1 : i;
}

export function gaussianBlur(gray: GrayImage, radius: number): GrayImage {
  if (!(radius > 0)) return gray;
  const { width, height, data } = gray;
  const k = gaussianKernel(radius);
  const r = (k.length - 1) / 2;
  const tmp = new Uint8Array(width * height);
  for (let y = 0; y < height; y += 1) {
    const row = y * width;
    for (let x = 0; x < width; x += 1) {
      let acc = 0;
      for (let i = -r; i <= r; i += 1)
        acc += (k[i + r] ?? 0) * (data[row + clampIndex(x + i, width)] ?? 0);
      tmp[row + x] = Math.min(255, Math.max(0, Math.round(acc)));
    }
  }
  const out = new Uint8Array(width * height);
  for (let x = 0; x < width; x += 1) {
    for (let y = 0; y < height; y += 1) {
      let acc = 0;
      for (let i = -r; i <= r; i += 1)
        acc += (k[i + r] ?? 0) * (tmp[clampIndex(y + i, height) * width + x] ?? 0);
      out[y * width + x] = Math.min(255, Math.max(0, Math.round(acc)));
    }
  }
  return { width, height, data: out };
}

export function minFilter(gray: GrayImage, size: number): GrayImage {
  if (!(size > 1)) return gray;
  const { width, height, data } = gray;
  const m = Math.floor(size / 2);
  const tmp = new Uint8Array(width * height);
  for (let y = 0; y < height; y += 1) {
    const row = y * width;
    for (let x = 0; x < width; x += 1) {
      let v = 255;
      for (let i = -m; i <= m; i += 1) {
        const p = data[row + clampIndex(x + i, width)] ?? 0;
        if (p < v) v = p;
      }
      tmp[row + x] = v;
    }
  }
  const out = new Uint8Array(width * height);
  for (let x = 0; x < width; x += 1) {
    for (let y = 0; y < height; y += 1) {
      let v = 255;
      for (let i = -m; i <= m; i += 1) {
        const p = tmp[clampIndex(y + i, height) * width + x] ?? 0;
        if (p < v) v = p;
      }
      out[y * width + x] = v;
    }
  }
  return { width, height, data: out };
}

export type UnsharpBand = {
  /** The row range [from, to) of the fitted image the mask applies to. */
  rows: [number, number];
  /** Percent, Pillow's unit: 180 means 180 percent. */
  amount: number;
  radius?: number;
  threshold?: number;
};

export function unsharpBand(gray: GrayImage, band: UnsharpBand): GrayImage {
  const radius = band.radius ?? 8;
  const threshold = band.threshold ?? 3;
  const blurred = gaussianBlur(gray, radius);
  const out = new Uint8Array(gray.data);
  const from = Math.max(0, band.rows[0]);
  const to = Math.min(gray.height, band.rows[1]);
  for (let y = from; y < to; y += 1) {
    for (let x = 0; x < gray.width; x += 1) {
      const i = y * gray.width + x;
      const v = gray.data[i] ?? 0;
      const diff = v - (blurred.data[i] ?? 0);
      if (Math.abs(diff) > threshold) {
        const s = Math.round(v + (diff * band.amount) / 100);
        out[i] = s < 0 ? 0 : s > 255 ? 255 : s;
      }
    }
  }
  return { width: gray.width, height: gray.height, data: out };
}

File: crates/turboslide-native/src/filters.rs
//! The optional spatial filters of the photograph pipeline, the Turboslide definitions pinned in
//! packages/effects/src/filters.ts (not Pillow's): a Gaussian blur with sigma = radius, kernel
//! radius ceil(3 sigma), weights exp(-x^2 / 2 sigma^2) normalized to one, separable, edges
//! replicate, each pass rounded with Math.round; a minimum filter over a size by size window;
//! and Pillow's unsharp rule applied to a band of rows.

use crate::image::Gray;
use crate::jsmath::js_round;

/// The normalized Gaussian taps, `exp` from libm so every target agrees with V8's fdlibm exp.
pub fn gaussian_kernel(sigma: f64) -> Vec<f64> {
    let radius = (3.0 * sigma).ceil() as i64;
    let len = (radius * 2 + 1) as usize;
    let mut k = vec![0f64; len];
    let mut sum = 0.0;
    for i in -radius..=radius {
        let w = libm::exp(-((i * i) as f64) / (2.0 * sigma * sigma));
        k[(i + radius) as usize] = w;
        sum += w;
    }
    for w in &mut k {
        *w /= sum;
    }
    k
}

fn clamp_index(i: i64, n: usize) -> usize {
    if i < 0 {
        0
    } else if i >= n as i64 {
        n - 1
    } else {
        i as usize
    }
}

fn round_clamp(acc: f64) -> u8 {
    let v = js_round(acc);
    if v < 0.0 {
        0
    } else if v > 255.0 {
        255
    } else {
        v as u8
    }
}

pub fn gaussian_blur(gray: &Gray, radius: f64) -> Gray {
    if !(radius > 0.0) {
        return gray.clone();
    }
    let (width, height) = (gray.width, gray.height);
    let k = gaussian_kernel(radius);
    let r = ((k.len() - 1) / 2) as i64;
    let mut tmp = vec![0u8; width * height];
    for y in 0..height {
        let row = y * width;
        for x in 0..width {
            let mut acc = 0.0;
            for i in -r..=r {
                acc +=
                    k[(i + r) as usize] * gray.data[row + clamp_index(x as i64 + i, width)] as f64;
            }
            tmp[row + x] = round_clamp(acc);
        }
    }
    let mut out = vec![0u8; width * height];
    for x in 0..width {
        for y in 0..height {
            let mut acc = 0.0;
            for i in -r..=r {
                acc +=
                    k[(i + r) as usize] * tmp[clamp_index(y as i64 + i, height) * width + x] as f64;
            }
            out[y * width + x] = round_clamp(acc);
        }
    }
    Gray {
        width,
        height,
        data: out,
    }
}

pub fn min_filter(gray: &Gray, size: f64) -> Gray {
    if !(size > 1.0) {
        return gray.clone();
    }
    let (width, height) = (gray.width, gray.height);
    let m = (size / 2.0).floor() as i64;
    let mut tmp = vec![0u8; width * height];
    for y in 0..height {
        let row = y * width;
        for x in 0..width {
            let mut v = 255u8;
            for i in -m..=m {
                let p = gray.data[row + clamp_index(x as i64 + i, width)];
                if p < v {
                    v = p;
                }
            }
            tmp[row + x] = v;
        }
    }
    let mut out = vec![0u8; width * height];
    for x in 0..width {
        for y in 0..height {
            let mut v = 255u8;
            for i in -m..=m {
                let p = tmp[clamp_index(y as i64 + i, height) * width + x];
                if p < v {
                    v = p;
                }
            }
            out[y * width + x] = v;
        }
    }
    Gray {
        width,
        height,
        data: out,
    }
}

/// The unsharp band: `out = in + (in - blur) * amount / 100` where |in - blur| > threshold, on
/// rows [from, to) of the fitted image only. `amount` is a percent, Pillow's unit.
pub fn unsharp_band(
    gray: &Gray,
    rows: [f64; 2],
    amount: f64,
    radius: Option<f64>,
    threshold: Option<f64>,
) -> Gray {
    let radius = radius.unwrap_or(8.0);
    let threshold = threshold.unwrap_or(3.0);
    let blurred = gaussian_blur(gray, radius);
    let mut out = gray.data.clone();
    let from = f64::max(0.0, rows[0]) as usize;
    let to = f64::min(gray.height as f64, rows[1]) as usize;
    for y in from..to {
        for x in 0..gray.width {
            let i = y * gray.width + x;
            let v = gray.data[i] as f64;
            let diff = v - blurred.data[i] as f64;
            if diff.abs() > threshold {
                let s = js_round(v + (diff * amount) / 100.0);
                out[i] = if s < 0.0 {
                    0
                } else if s > 255.0 {
                    255
                } else {
                    s as u8
                };
            }
        }
    }
    Gray {
        width: gray.width,
        height: gray.height,
        data: out,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kernel_sums_to_one() {
        for sigma in [0.4, 0.5, 0.6, 2.0, 8.0] {
            let k = gaussian_kernel(sigma);
            assert_eq!(k.len(), ((3.0 * sigma).ceil() as usize) * 2 + 1);
            let sum: f64 = k.iter().sum();
            assert!((sum - 1.0).abs() < 1e-12, "sigma {sigma} sums to {sum}");
        }
    }

    #[test]
    fn blur_keeps_a_flat_field() {
        let g = Gray {
            width: 5,
            height: 3,
            data: vec![90; 15],
        };
        assert_eq!(gaussian_blur(&g, 1.0).data, vec![90; 15]);
        assert_eq!(gaussian_blur(&g, 0.0).data, vec![90; 15]);
    }

    #[test]
    fn min_filter_takes_the_window_minimum() {
        let g = Gray {
            width: 3,
            height: 1,
            data: vec![10, 200, 50],
        };
        assert_eq!(min_filter(&g, 3.0).data, vec![10, 10, 50]);
        assert_eq!(min_filter(&g, 1.0).data, vec![10, 200, 50]);
    }
}
